File: java/src/main/java/io/github/muthuishere/toolnexus/NativeTool.java
package io.github.muthuishere.toolnexus;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.function.BiFunction;
import java.util.function.Function;

/**
 * Programmatic native tool (source {@code "native"}). Wrap a plain function as a
 * uniform {@link Tool}.
 *
 * <p>The function may return a {@link String} (wrapped as output) or a full
 * {@link ToolResult}. A thrown exception becomes an error result.
 */
public final class NativeTool implements Tool {
    private static final Map<String, Object> EMPTY_SCHEMA = emptySchema();

    private final String name;
    private final String description;
    private final Map<String, Object> inputSchema;
    private final String source;
    private final BiFunction<Map<String, Object>, ToolContext, Object> fn;

    private NativeTool(String name, String description, Map<String, Object> inputSchema, String source,
                       BiFunction<Map<String, Object>, ToolContext, Object> fn) {
        this.name = name;
        this.description = description;
        this.inputSchema = inputSchema == null ? EMPTY_SCHEMA : inputSchema;
        this.source = source == null ? "native" : source;
        this.fn = fn;
    }

    /** Build from a function that ignores the context. */
    public static NativeTool of(String name, String description, Map<String, Object> inputSchema,
                                Function<Map<String, Object>, Object> fn) {
        return new NativeTool(name, description, inputSchema, "native", (args, ctx) -> fn.apply(args));
    }

    /** Build from a function that receives the context. */
    public static NativeTool of(String name, String description, Map<String, Object> inputSchema,
                                BiFunction<Map<String, Object>, ToolContext, Object> fn) {
        return new NativeTool(name, description, inputSchema, "native", fn);
    }

    @Override
    public String name() {
        return name;
    }

    @Override
    public String description() {
        return description;
    }

    @Override
    public Map<String, Object> inputSchema() {
        return inputSchema;
    }

    @Override
    public String source() {
        return source;
    }

    @Override
    public ToolResult execute(Map<String, Object> args, ToolContext ctx) {
        try {
            Object out = fn.apply(args == null ? new LinkedHashMap<>() : args, ctx);
            if (out instanceof ToolResult) {
                return (ToolResult) out;
            }
            if (out instanceof String) {
                return ToolResult.ok((String) out);
            }
            return ToolResult.ok(Json.stringify(out));
        } catch (Exception e) {
            Throwable cause = e.getCause() != null ? e.getCause() : e;
            String msg = cause.getMessage() == null ? String.valueOf(cause) : cause.getMessage();
            return ToolResult.error(msg);
        }
    }

    private static Map<String, Object> emptySchema() {
        Map<String, Object> s = new LinkedHashMap<>();
        s.put("type", "object");
        s.put("properties", new LinkedHashMap<>());
        s.put("additionalProperties", false);
        return s;
    }
}

File: csharp/src/Toolnexus/NativeTool.cs
namespace Toolnexus;

/// <summary>
/// Programmatic native tool (source <c>"native"</c>). Wrap a plain function as a
/// uniform <see cref="ITool"/>.
///
/// <para>The function may return a <see cref="string"/> (wrapped as output) or a full
/// <see cref="ToolResult"/>. A thrown exception becomes an error result.</para>
/// </summary>
public sealed class NativeTool : ITool
{
    private readonly Func<IDictionary<string, object?>, ToolContext?, Task<object?>> _fn;

    public string Name { get; }
    public string Description { get; }
    public IDictionary<string, object?> InputSchema { get; }
    public string Source { get; }

    private NativeTool(string name, string description, IDictionary<string, object?>? inputSchema,
        Func<IDictionary<string, object?>, ToolContext?, Task<object?>> fn)
    {
        Name = name;
        Description = description;
        InputSchema = inputSchema ?? EmptySchema();
        Source = "native";
        _fn = fn;
    }

    /// <summary>Build from a synchronous function that ignores the context.</summary>
    public static NativeTool Of(string name, string description, IDictionary<string, object?>? inputSchema,
        Func<IDictionary<string, object?>, object?> fn)
        => new(name, description, inputSchema, (args, _) => Task.FromResult(fn(args)));

    /// <summary>Build from a synchronous function that receives the context.</summary>
    public static NativeTool Of(string name, string description, IDictionary<string, object?>? inputSchema,
        Func<IDictionary<string, object?>, ToolContext?, object?> fn)
        => new(name, description, inputSchema, (args, ctx) => Task.FromResult(fn(args, ctx)));

    /// <summary>Build from an asynchronous function.</summary>
    public static NativeTool OfAsync(string name, string description, IDictionary<string, object?>? inputSchema,
        Func<IDictionary<string, object?>, ToolContext?, Task<object?>> fn)
        => new(name, description, inputSchema, fn);

    public async Task<ToolResult> ExecuteAsync(IDictionary<string, object?> args, ToolContext? ctx = null)
    {
        try
        {
            var output = await _fn(args ?? new Dictionary<string, object?>(), ctx).ConfigureAwait(false);
            return output switch
            {
                ToolResult tr => tr,
                string s => ToolResult.Ok(s),
                null => ToolResult.Ok(""),
                _ => ToolResult.Ok(Json.Stringify(output)),
            };
        }
        catch (Exception e)
        {
            var cause = e.InnerException ?? e;
            return ToolResult.Error(cause.Message ?? cause.ToString());
        }
    }

    internal static Dictionary<string, object?> EmptySchema() => new()
    {
        ["type"] = "object",
        ["properties"] = new Dictionary<string, object?>(),
        ["additionalProperties"] = false,
    };
}
